### ISSUE TITLE
process ml-1m dataset

### ISSUE THREAD

**Issue by ZuoxiYang** [pmixer/SASRec.pytorch#3]
Hi, many thanks for your amazing work. When I checked the ml-1m dataset, I found that it has been already processed, could you share the methods about how you process this dataset (or the mapping relation between the original user/item id and mapped user/item id)?

**Comment by pmixer** [comment]
> Hi, many thanks for your amazing work. When I checked the ml-1m dataset, I found that it has been already processed, could you share the methods about how you process this dataset (or the mapping relation between the original user/item id and mapped user/item id)?

Hi Yang, thx for your comment, the preprocessed datasets are borrowed from original tensorflow implementation, here's a sample of data preprocessing code in their repo: https://github.com/kang205/SASRec/blob/master/data/DataProcessing.py and here's the discussion about the ml-1m preprocessing issue. It could be also be easy done by writing our own preprocessing script as we just need to sort the ml-1m records by user_id as major key and then timestamp as second key, then re-assign id to users to get the pre-processed data.

**Comment by ZuoxiYang** [comment]
> > Hi, many thanks for your amazing work. When I checked the ml-1m dataset, I found that it has been already processed, could you share the methods about how you process this dataset (or the mapping relation between the original user/item id and mapped user/item id)?
> 
> Hi Yang, thx for your comment, the preprocessed datasets are borrowed from original tensorflow implementation, here's a sample of data preprocessing code in their repo: https://github.com/kang205/SASRec/blob/master/data/DataProcessing.py and here's the discussion about the ml-1m preprocessing issue. It could be also be easy done by writing our own preprocessing script as we just need to sort the ml-1m records by user_id as major key and then timestamp as second key, then re-assign id to users to get the pre-processed data.

Yeah, because some other author would remove some cold users/items, I do not know if you do the same way, so I asked this question. So in the ml-1m dataset, you just sort the records by timestamp?

**Comment by pmixer** [comment]
> > > Hi, many thanks for your amazing work. When I checked the ml-1m dataset, I found that it has been already processed, could you share the methods about how you process this dataset (or the mapping relation between the original user/item id and mapped user/item id)?
> > 
> > 
> > Hi Yang, thx for your comment, the preprocessed datasets are borrowed from original tensorflow implementation, here's a sample of data preprocessing code in their repo: https://github.com/kang205/SASRec/blob/master/data/DataProcessing.py and here's the discussion about the ml-1m preprocessing issue. It could be also be easy done by writing our own preprocessing script as we just need to sort the ml-1m records by user_id as major key and then timestamp as second key, then re-assign id to users to get the pre-processed data.
> 
> Yeah, because some other author would remove some cold users/items, I do not know if you do the same way, so I asked this question. So in the ml-1m dataset, you just sort the records by timestamp?

Actually I reused a lot of components of original tf implementation, including the pre-processed datasets, they seems performed code users/items(with less than 3 appear time) in train/va/test set generation rather than in raw data preprocessing, pls see https://github.com/kang205/SASRec/blob/d0b823a1b2845ea90b090eaf6d3386810005b83a/util.py#L27 , but your question is reasonable and quite good, as original ml-1m has 1000209 ratings but the dataset used currently only has 999611 rows, maybe due to duplicates etc. I strongly encourge trying your own preprocessing scripts for ml-1m raw data if its a big concern for you(yeah, just sort the ratings by (user, timestamp), re-assign ids to users should just be enough)  personally I do have the plan to try my own data preprocessing scripts to validate it.

**Comment by ZuoxiYang** [comment]
> > > > Hi, many thanks for your amazing work. When I checked the ml-1m dataset, I found that it has been already processed, could you share the methods about how you process this dataset (or the mapping relation between the original user/item id and mapped user/item id)?
> > > 
> > > 
> > > Hi Yang, thx for your comment, the preprocessed datasets are borrowed from original tensorflow implementation, here's a sample of data preprocessing code in their repo: https://github.com/kang205/SASRec/blob/master/data/DataProcessing.py and here's the discussion about the ml-1m preprocessing issue. It could be also be easy done by writing our own preprocessing script as we just need to sort the ml-1m records by user_id as major key and then timestamp as second key, then re-assign id to users to get the pre-processed data.
> > 
> > 
> > Yeah, because some other author would remove some cold users/items, I do not know if you do the same way, so I asked this question. So in the ml-1m dataset, you just sort the records by timestamp?
> 
> Actually I reused a lot of components of original tf implementation, including the pre-processed datasets, they seems performed code users/items(with less than 3 appear time) in train/va/test set generation rather than in raw data preprocessing, pls see https://github.com/kang205/SASRec/blob/d0b823a1b2845ea90b090eaf6d3386810005b83a/util.py#L27 , but your question is reasonable and quite good, as original ml-1m has 1000209 ratings but the dataset used currently only has 999611 rows, maybe due to duplicates etc. I strongly encourge trying your own preprocessing scripts for ml-1m raw data if its a big concern for you(yeah, just sort the ratings by (user, timestamp), re-assign ids to users should just be enough) personally I do have the plan to try my own data preprocessing scripts to validate it.

And after I ran your codes, its results were slightly below the results of the paper, so are "--maxlen=200 --dropout_rate=0.2"  the optimal hyper-parameters? Could you send me the optimal hyper-parameters? Or maybe there are any other reasons?

**Comment by pmixer** [comment]
> parameters

Pls refer to the original paper for optimal hyper-parameters, I'm not the paper author, just reproducing the work in PyTorch for personal use. And its also a good point to mention that results were slightly below the tf implementation, it takes more epochs to achieve comparable performance with original tf implementation which is reasonable as tf implemention's multi-head attention layer suffers (bit) leaky attention weights issue. I would really appreciate it if you could find some way to make it coverage faster.

**Comment by ZuoxiYang** [comment]
> > parameters
> 
> Pls refer to the original paper for optimal hyper-parameters, I'm not the paper author, just reproducing the work in PyTorch for personal use. And its also a good point to mention that results were slightly below the tf implementation, it takes more epochs to achieve comparable performance with original tf implementation which is reasonable as tf implemention's multi-head attention layer suffers (bit) leaky attention weights issue. I would really appreciate it if you could find some way to make it coverage faster.

I checked the hyper-parameters again, I found that in your codes, the number of negative-sample is "maxlen" (which was set 200), while the number of negative-sample is 100 reported in the paper, so maybe it is a reason that causes the different results?

**Comment by pmixer** [comment]
@ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function: https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135 the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).

**Comment by ZuoxiYang** [comment]
> @ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function:
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135
> 
> the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).
Thanks! I agree with your opinion. I also have changed the iterations as 1001, then the results were improved slightly. But it always slightly below the results of the paper. I will take more experiments.

**Comment by ZuoxiYang** [comment]
> @ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function:
> 
> https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135
> 
> the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).

And in your codes, the negative sample would be used for both model training and evaluation?

**Comment by alan-ai-learner** [comment]
> > @ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function:
> > https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135
> > 
> > the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).
> > Thanks! I agree with your opinion. I also have changed the iterations as 1001, then the results were improved slightly. But it always slightly below the results of the paper. I will take more experiments.

Hi @ZuoxiYang , have you achieve the similar results?
If yes, can you share you parameters.

Thanks!

**Comment by pmixer** [comment]
> > > @ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function:
> > > https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135
> > > 
> > > the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).
> > > Thanks! I agree with your opinion. I also have changed the iterations as 1001, then the results were improved slightly. But it always slightly below the results of the paper. I will take more experiments.
> 
> Hi @ZuoxiYang , have you achieve the similar results? If yes, can you share you parameters.
> 
> Thanks!

not sure whether he/she did, it's way too far from now.

the perf issue should have been resolved by changing weight init method https://github.com/pmixer/SASRec.pytorch/commit/1d8aaa2dcf3837b0518692faee8af577cc166188

though the results would still be slightly different for each run, due to randomness in weight initialization and negative sampling.

**Comment by pmixer** [comment]
> > @ZuoxiYang 200 is the seq length, 100 negative sample number is hard-coded in evaluate function:
> > https://github.com/pmixer/SASRec.pytorch/blob/4a853a6f956ca4ebcd802df45a9d5ee7c291d3d5/utils.py#L135
> > 
> > the pytorch version converges slower than original version although I tried to keep the details as consistent as possible, pls just train more iterations(I trained 601 epochs rather than 201 by default, current assumption is after causality issue has been fixed the model should converge slower as expected, meanwhile, as negative samples are draw randomly, there's inherent uncertainty in implementation, we can not expect exactly same results as reported in the paper).
> 
> And in your codes, the negative sample would be used for both model training and evaluation?

not sure what "the negative sample would be used for both model training and evaluation" means, negative sampling is a common approach in recsys research work, if want more stable results, just rank all items like in https://github.com/pmixer/TiSASRec.debug
